RecoLocalCalo/HcalRecProducers/src/HcalESProducerGPU.h

```diff
@@ -0,0 +1,166 @@
+#ifndef RecoLocalCalo_HcalRecProducers_src_HcalESProducerGPU_h
+#define RecoLocalCalo_HcalRecProducers_src_HcalESProducerGPU_h
+
+#include "FWCore/Framework/interface/ESProducer.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Utilities/interface/typelookup.h"
+#include "FWCore/Framework/interface/eventsetuprecord_registration_macro.h"
+#include "FWCore/Framework/interface/ESTransientHandle.h"
+#include "FWCore/Framework/interface/ModuleFactory.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "FWCore/Framework/interface/ESProductHost.h"
+#include "FWCore/Utilities/interface/ReusableObjectHolder.h"
+
+#include <iostream>
+#include <array>
+#include <tuple>
+#include <utility>
+
+template <typename Record, typename Target, typename Source>
+class HcalESProducerGPU : public edm::ESProducer {
+public:
+  explicit HcalESProducerGPU(edm::ParameterSet const& ps) {
+    auto const label = ps.getParameter<std::string>("label");
+    std::string name = ps.getParameter<std::string>("ComponentName");
+    auto cc = setWhatProduced(this, name);
+
+    cc.setConsumes(token_, edm::ESInputTag{"", label});
+  }
+
+  std::unique_ptr<Target> produce(Record const& record) {
+    // retrieve conditions in old format
+    auto sourceProduct = record.getTransientHandle(token_);
+
+    return std::make_unique<Target>(*sourceProduct);
+  }
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& confDesc) {
+    edm::ParameterSetDescription desc;
+
+    std::string label = Target::name() + "ESProducer";
+    desc.add<std::string>("ComponentName", "");
+    desc.add<std::string>("label", "")->setComment("Product Label");
+    confDesc.add(label, desc);
```

Comment:
not sure. As this is a template, what string will sit in the default label?

Reply:
I think it will pick the name from what is passed to `DEFINE_FWK_EVENTSETUP_MODULE`.
So
```
using HcalGainsGPUESProducer = HcalESProducerGPU<HcalGainsRcd, HcalGainsGPU, HcalGains>;
DEFINE_FWK_EVENTSETUP_MODULE(HcalGainsGPUESProducer);
```
should get a default label of `hcalGainsGPUESProducer`.

Reply:
should be fine then i think, thx

Reply:
Correct, the label comes from the class name passed to `DEFINE_FWK_EVENTSETUP_MODULE` (or `DEFINE_FWK_MODULE` for ED).